src/Run.purs

```diff
@@ -104,42 +105,80 @@ send
   → Run r a
 send = RunM <<< liftF
 
+-- | Casts some set of effects to a wider set of effects via a left-biased
+-- | union.
+expand
+  ∷ ∀ r1 r2 rx a
+  . Union r1 rx r2
+  ⇒ Run r1 a
+  → Run r2 a
+expand = unsafeCoerce
+
 -- | Extracts the value from a fully interpreted program.
 run ∷ ∀ a. Run () a → a
 run = unwrap >>> runFree \_ → unsafeCrashWith "Control.Monad.Run: the impossible happened"
 
--- | Extracts the value from a program with only base effects.
+unBase ∷ ∀ f. VariantF (base ∷ FProxy f) ~> f
+unBase = case_ # on (SProxy ∷ SProxy "base") id
+
+-- | Extracts the value from a program with only base effects. This assumes
+-- | stack safety under Monadic recursion.
 runBase
   ∷ ∀ f a
+  . Monad f
+  ⇒ Run (base ∷ FProxy f) a
+  → f a
+runBase = unwrap >>> loop
+  where
+  loop = resume' (\b f → f <$> unBase b >>= loop) pure
+
+-- | Extracts the value from a program with only base effects using `MonadRec`
+-- | to preserve stack safety.
+runBaseRec
+  ∷ ∀ f a
   . MonadRec f
   ⇒ Run (base ∷ FProxy f) a
   → f a
-runBase = unwrap >>> foldFree go
+runBaseRec = unwrap >>> foldFree (case_ # on (SProxy ∷ SProxy "base") id)
+
+-- | Interprets the base effect into some Monad `m` via continuation passing.
+runBaseWithCont
+  ∷ ∀ f m a b
+  . Monad m
+  ⇒ (f (Unit → m b) → m b)
+  → (a → m b)
+  → Run (base ∷ FProxy f) a
+  → m b
+runBaseWithCont k1 k2 = unwrap >>> loop
   where
-  go ∷ VariantF (base ∷ FProxy f) ~> f
-  go = case_ # on (SProxy ∷ SProxy "base") id
+  loop = resume' (\b f → k1 (unBase $ (\b' _ → loop (f b')) <$> b)) k2
 
 -- | Interprets an effect functor into the base effect, eliminating the label.
-interpret
-  ∷ ∀ sym f m r1 r2 r3 a
-  . RowCons sym (FProxy f) r2 r1
-  ⇒ RowCons "base" (FProxy m) r2 r3
+runWithBase
+  ∷ ∀ sym f m r1 r2 a
+  . RowCons sym (FProxy f) (base ∷ FProxy m | r2) r1
   ⇒ IsSymbol sym
   ⇒ Functor m
   ⇒ SProxy sym
-  → (f ~> m)
+  → (f (Run r1 a) → m (Run r1 a))
```

Comment:
Should this be:

```PureScript
(forall a r. f (Run r a) -> m (Run r a))
```
?

Or, do we want to allow for people changing values here?

Reply:
Possibly. I considered it, but I also didn't want to rule out use cases. The rank-n thing is nice because it restricts the possible implementation, so it's harder to get it wrong. But I also wondered if there's a use case in being able to rewrite the rest of the computation. With this formulation an interpreter can take the tail and derive something entirely new from it.

Reply:
I'm undecided though.

Reply:
Yeah, I dunno either. I think if I wanted to do something like that, I probably want the ability to change the type as well. Something like `(f (Run r a) -> m (Run r b))` or whatever. Otherwise, I'd probably just want the natural transformation-y version.

I hate to suggest _more_ functions, but you could build the more natural transformation-y ones on top of these two. Maybe move these two to an "experimental" module or something?

Reply:
It's not possible to change the type at each turn of the interpreter and still know what the input type is. One advantage of the more permissive type is that there are no inference issues due to let generalization. I only say that because writing out all the effects again with one missing, is annoying. So I think it should either by fully permissive or fully restrictive with a nat.

Reply:
But yeah, there's no reason both can't exist. Some of the other libraries have `someFooNat` variations. I think I agree that the natural transformations should probably be the default.

Reply:
I don't really like primes in APIs, but maybe just having `runWithBase'` be the more permissive one.